### PR TITLE
Generic AST traversal

### PR DESCRIPTION
**Pull request by jgberry**
There are quite a few changes here so I'll try to break them down as best I can:

### Generally simplified the `_bindings` and `_requirements` modules

The `_bindings` and `_requirements` modules both perform AST traversal so they share a lot of behavior. By extracting the common functionality of both modules to a new module, named `_visitor`, I was able to significantly reduce the size of both modules.

### Improved overall performance

My initial intention was to utilize existing functionality from the standard library `ast` module to simplify the AST traversal done in both the `_bindings` and `_requirements` modules. However, after running some primitive benchmarks and realizing an over 20% reduction in performance I decided to scrap the idea. Instead, I put together the `_visitors` module and implemented a high speed alternative to the `ast.NodeVisitor` class. Swapping out the previously used `functools.singledispatch` AST traversal approach for the `_visitor.NodeVisitor` resulted in performance improvements in the 10%-15% range.

### Fixes

While making the above changes I ran into a few existing issues which I fixed:

1. Fixed bindings not being discovered in return statements ([test case](https://github.com/bwhmather/ssort/commit/a1f4b6f6e5c757cffd50348a1bfd7d7461d828d2)).
2. Fixed bindings not being discovered in exception handler type expressions ([test case](https://github.com/bwhmather/ssort/commit/7ca8ea2ba590a3a2084c5a6893baa1933175b3d0)).
3. Fixed dict not resolving requirements from its keys ([test case](https://github.com/bwhmather/ssort/commit/7517853bae7530dfa79d0e2fce18109440b0abe7))
4. Fixed augmented assignment not resolving requirements from its target ([test case](https://github.com/bwhmather/ssort/commit/b3c5ab5e208421cdf227ea5eb51652a0d2c9211b))
5. Fixed annotated assignment not resolving requirements from its annotations ([test case](https://github.com/bwhmather/ssort/commit/ee462bd641d9f9ba2d3ddcb617331d6c0f7ee7a7))
6. Fixed functions and lambdas not resolving requirements from their arguments and annotations ([test case](https://github.com/bwhmather/ssort/commit/721f5c7633136376a5dcb5caaf6ba897d0046b0e))
7. Enabled walrus tests for Python 3.8 ([test case](https://github.com/bwhmather/ssort/commit/f53d05f501ff2149eb6495311571f90d552e5e65))
8. Fixed with statement not resolving requirements from its optional_vars ([test case](https://github.com/bwhmather/ssort/pull/42/commits/74c48ca4102ede04bfdaccf8166c14e0f80d452d))
9. Fixed for statement not resolving requirements from its target ([test case](https://github.com/bwhmather/ssort/pull/42/commits/e5aba6d438e7c79cb4cd8c4a7b6f79c55d9e2e7c))

### Type annotations

Additionally, I added type annotations to the `_bindings`, `_requirements`, and `_visitor` modules.

### CODE REVIEW

**Comment by bwhmather** [comment]
First things first:  this is fantastic work, thank you.

Regarding your list of bugs: did you have a shortcut to finding them, or was it a case of just putting in the work, reading the code carefully and writing the tests?

Regarding performance: thanks for checking.  The hit for ast.walk was not as bad as I had feared, but I think this is a better solution.

One thing that is interesting to me is the choice between visiting in source code order and evaluation order.  The requirements code and most of the bindings code were written at different times (grrrrrr walrus operator) and it looks like I made a different call with each of them.  It should make zero difference to the final sort order, so I'm not going to overthink it.

File: src/ssort/_bindings.py [hunk 1]
```diff
@@ -1,802 +1,76 @@
...
+        yield from node.names
+
+    @register_visitor(ast.Lambda)
+    def visit_lambda(self, node: ast.Lambda) -> Iterable[str]:
+        yield from self.visit(node.args)
+
+    @register_visitor(ast.Name)
+    def visit_name(self, node: ast.Name) -> Iterable[str]:
+        if isinstance(node.ctx, ast.Store):
+            yield node.id
```
Comment (bwhmather) [inline src/ssort/_bindings.py:60]: I like that this just handles everything I was attempting to do with `_flatten_target`.  Branching later is good.

**Review by bwhmather — COMMENTED** [review]
I think your earlier approach with `singledispatch` and `ast.iter_child_nodes` was cleaner.  It makes what you are doing with each child node more explict (`get_bindings`/`get_requirements` vs `self.visit`), and I suspect that it will generalise better to `_method_requirements.py`.  Can your `register_visitor`/`_find_visitor_methods` combo be merged to keep the 10% to 15% speedup over `singledispatch` with functions rather than methods, or do you think the speedup comes from somewhere else?

**Comment by jgberry** [comment]
> Regarding your list of bugs: did you have a shortcut to finding them, or was it a case of just putting in the work, reading the code carefully and writing the tests?

I think most, if not all, of the bugs I found where simple cases where a specific field on a specific type of AST node was ignored or not traversed as deep within the AST as it should have been. Take the [function and lambda arguments fix](https://github.com/bwhmather/ssort/pull/42/commits/da91607f5a44e03eb35be4ec958b92257a7d1a4c) for example, by comparing the implementation of `_get_requirements_for_function_def` to the signature of `ast.FunctionDef` it was fairly apparent that the implementation missed the traversal of the `ast.FunctionDef.args` field. From there, I wrote a test case which contained requirements in a functions arguments and validated that the currently implementation did not properly handle that case. Repeat that process for all `_get_bindings_for_*` and `_get_requirements_for_*` functions and I was able to snuff out quite a few bugs.

> I think your earlier approach with singledispatch and ast.iter_child_nodes was cleaner. It makes what you are doing with each child node more explict (get_bindings/get_requirements vs self.visit), and I suspect that it will generalise better to _method_requirements.py. Can your register_visitor/_find_visitor_methods combo be merged to keep the 10% to 15% speedup over singledispatch with functions rather than methods, or do you think the speedup comes from somewhere else?

Using `ast.iter_fields` was certainly the cleaner approach (and I just  realized it could be even cleaner if I used `ast.iter_child_nodes` instead of `ast.iter_fields`) as it completely avoided all the boilerplate in the `_visitor` module. However, I felt the loss in performance wasn't acceptable. As for the `_visitor` module itself, I believe most of the speed up comes from avoiding `functools.singledispatch` and instead using a single dictionary lookup in it's place. I don't think it would be too difficult to keep the `_visitor` module functionality but expose it through an interface similar to `functools.singledispatch` to maintain that cleanliness you are referring to. I'll try playing around with it.

File: src/ssort/_requirements.py [hunk 2]
```diff
@@ -23,15 +25,11 @@ class Requirement:
...
 
 
-@functools.singledispatch
-def get_requirements(node):
-    raise NotImplementedError(
-        f"could not find requirements for unsupported node:  {node!r}"
-        f"at line {node.lineno}, column: {node.col_offset}"
-    )
+_requirements_visitor: NodeVisitor[Requirement] = NodeVisitor()
+get_requirements = _requirements_visitor.visit
```
Comment (bwhmather) [inline src/ssort/_requirements.py:29]: This is an improvement, but I'm concerned that there are now two modes of recursion:
  1. A `_get_requirements` function recurses by calling `get_requirements`.
  2. A visitor function calls `visitor.visit`.

I think there are two options:
  1. Move recursion entirely into the visitor:  Allow get bindings to yield only the bindings that are bound directly by the visited node.
  2. Move recursion entirely into `get_requirements`:  Make the visitor just yield immediate child nodes and `for child in visit(node): yield from get_requirements(child)` as the default implementation.

My vote is with 2, as I think people will always want to read the get_requirements/get_binding functions first, and it makes it a bit easier to stop recursion if necessary.

Comment (bwhmather) [reply src/ssort/_requirements.py:29]: Perhaps the other way to frame this is in terms of composition vs inheritance.  Effectively a NodeVisitor instance is a kind of class that subclasses the basic implementation.  I'm struggling to come up with a coherent argument from there, but that's the instinct that this is triggering.

**Comment by bwhmather** [comment]
> Repeat that process for all _get_bindings_for_* and _get_requirements_for_* functions and I was able to snuff out quite a few bugs

That sounds an awful lot like putting in the work.  I'm very grateful that someone did it.

File: src/ssort/_requirements.py (hunk 2 above)
Comment (jgberry) [reply src/ssort/_requirements.py:29]: Hmm, that does seem a lot cleaner. Perhaps what we want is to re-purpose the `_visitor` module to provide a high speed `iter_child_nodes` function which parallels the `ast.iter_child_node` function. I think that would enforce a cleaner separation between the `_visitor` module and the `_bindings` and `_requirements` modules.

**Comment by jgberry** [comment]
> Move recursion entirely into get_requirements: Make the visitor just yield immediate child nodes and for child in visit(node): yield from get_requirements(child) as the default implementation.

Ok, I cleaned things up to align with the above specification; however, it's was difficult to do so without taking a significant performance hit. [f6a1e2c](https://github.com/bwhmather/ssort/pull/42/commits/f6a1e2c049745dcecb5466ade591ce780d3655d3) has performance roughly equivalent to master so we lost the 15% speed up.

Ultimately, it seems this comes down to a trade off between cleanliness and performance.

**Comment by bwhmather** [comment]
Performance requirements are basically arbitrary.  I think you have the right balance here.

I'm guessing from the lack of fanfare that this doesn't give us 3.10 support straight off, but I don't see the gap.  Does `isinstance(node.ctx, ast.Store)` not cover bindings in the match?

In any case, this seems like the right place to make the cut.  Will merge and make a new release to get your bug fixes out.

Fantastic work.

**Comment by bwhmather** [comment]
Released as version 0.11.1.  Thank you very much.

**Comment by jgberry** [comment]
> I'm guessing from the lack of fanfare that this doesn't give us 3.10 support straight off, but I don't see the gap. Does `isinstance(node.ctx, ast.Store)` not cover bindings in the match?

This, for the most part, should provide 3.10 support; however, there are no 3.10 unit tests. I’m guessing there may be some edge cases with match statements which aren’t fully captured by these changes.

Next steps are to simplify the _method_requirements module in a similar fashion to the _bindings and _requirements modules and then to work on 3.10 test coverage.
